File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.py
class DSU:
    def __init__(self, n: int) -> None:
        self.parent = [u for u in range(n + 1)]
        self.size = n
        self.group = n
    
    def find(self, u: int) -> int:
        if u != self.parent[u]:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]
    
    def union(self, u: int, v: int) -> bool:
        u = self.find(u)
        v = self.find(v)
        if u != v:
            self.parent[u] = v
            self.group -= 1
            return True
        return False

class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        Alice, Bob = DSU(n), DSU(n)
        connect_edge = 0
        
        for edge in edges:
            type_path, u, v = edge
            if type_path == 3:
                connect_edge += int(Alice.union(u, v) | Bob.union(u, v))
        
        for edge in edges:
            type_path, u, v = edge
            if type_path == 1:
                connect_edge += Alice.union(u, v)
            elif type_path == 2:
                connect_edge += Bob.union(u, v)
                
        if Alice.group == Bob.group == 1:
            return len(edges) - connect_edge
        
        return -1 

File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.cpp
class DSU {
private:
    vector<int> parent;
    int group;
    
public:
    DSU() {
        
    }
    
    DSU(int n) {
        parent.resize(n);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
        this->group = n - 1;
    }
    
    int find(int u) {
        if (u != parent[u]) {
            parent[u] = find(parent[u]);
        }
        return parent[u];
    }
    
    bool _union(int u, int v) {
        int pu = find(u);
        int pv = find(v);
        
        if (pu == pv) {
            return false;
        }
        
        parent[pu] = pv;
        group--;
        
        return true;
    }
    
    int get_group() {
        return group;
    }
};

class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        DSU Alice(n + 1), Bob(n + 1);
        int connect_edge = 0;
        int m = edges.size();
        
        for (int i = 0; i < m; ++i) {
            int t = edges[i][0];
            int u = edges[i][1], v = edges[i][2];
            
            if (t == 3) {
                connect_edge += (Alice._union(u, v) | Bob._union(u, v));
            }
        }
        
        for (int i = 0; i < m; ++i) {
            int t = edges[i][0];
            int u = edges[i][1], v = edges[i][2];
            
            if (t == 2) {
                connect_edge += Bob._union(u, v);
            }
            else if (t == 1) {
                connect_edge += Alice._union(u, v);
            }
        }
        
        if (Alice.get_group() == 1 and Bob.get_group() == 1) {
            return m - connect_edge;
        }
        
        return -1;
    }
};